gen_binary.py

```diff
@@ -0,0 +1,66 @@
+from misc.viz_utils import visualize_instances_dict
+import numpy as np
+import json
+from skimage import io
+import cv2
+import zarr
+import sys, getopt, os
+import slideio
+
+'''
+Converts all json files in a folder to binary zarr.
+The corresponding svs files also need to be present to obtain height and width.
+All 3 files will have the same name.
+Sample call: python gen_binary.py -i /mnt/ibm_lg/snigdha/svs -o /mnt/ibm_lg/snigdha/svs
+'''
+def main(argv):
+	path = ''
+	output_path = ''
+	try:
+		opts, args = getopt.getopt(argv,"i:o:h:w:",["in_path=","out_path="])
+	except getopt.GetoptError:
+		print('gen_binary.py -i <in_path> -o <output_path>')
+		sys.exit(2)
+
+	for opt, arg in opts:
+		if opt in ("-i", "--in_path"):
+			path = arg
+		elif opt in ("-o", "--out_path"):
+			output_path = arg
+
+	for file in os.listdir(path):
+		# processing each json file
+		if file.endswith(".json"):
+			inputfile = os.path.join(path,file)
+			# no need to run the process if the binary zarr already exists
+			if not os.path.exists(inputfile.split('.')[0]+".zarr"):
+				f = open(inputfile)
+				print("Loading json file ", inputfile)
+				data = json.load(f)
+				type_info = {"0" : ["nolabe", [255,255,255]]}
+				print("Extracting contour coordinates")
+				tile_info_dict = {}
+				for i in data['nuc']:
+					contour = data['nuc'][str(i)]['contour']
+					x_s = [a_tuple[0] for a_tuple in contour]
+					y_s = [a_tuple[1] for a_tuple in contour]
+					coords = np.empty((len(x_s),2))
+					coords[:,0] = x_s
+					coords[:,1] = y_s
+					tile_info_dict[i] = {'contour': np.int0(coords), 'type':'0'}
+
+				print("Obtaining image size from svs file")
+				svs = slideio.open_slide(inputfile.split('.')[0]+".svs",'SVS')
+				dims = svs.get_scene(0).rect
+				# shape of the max magnification level in svs.
+				# This is the image that will be overlaid so it's just an empty black image
+				r_mask = np.zeros((dims[3], dims[2]), np.uint8)
+				print("Obtaining overlaid output")
+				binary_image = visualize_instances_dict(r_mask, tile_info_dict,
```

Comment:
@snigdhaAgarwal I just checked the visualize_instances_dict function to see what they did differently and your hypothesis on using the draw_dot is completely incorrect, the draw_dot they use is for drawing a centroid inside the contour. This flag to draw the centroid is always set to False by default. - https://github.com/vqdang/hover_net/blob/be8ae2d621bfbddefd97591ef9df39252e108df9/misc/viz_utils.py#L121

They were still using cv2.drawContours and getting better results - https://github.com/vqdang/hover_net/blob/be8ae2d621bfbddefd97591ef9df39252e108df9/misc/viz_utils.py#L119

Have you used opencv function as well to fill the contours or was it skimage? could it be the difference in library. If not, please try to figure out what else could be the reason and look for differences in your code and try to understand the image processing differences behind it. Thank you!

Reply:
We did not try cv2 because we were worried about pyqt dependencies. I believe it is the difference in the libraries.

Reply:
Please try to confirm if that's the case. what does skimage does differently in the backend, it's a useful practise task to understand the differences between different functions

Reply:
Based on the C++ code of drawContours here: https://github.com/opencv/opencv/blob/68d15fc62edad980f1ffa15ee478438335f39cc3/modules/imgproc/src/drawing.cpp and comparing that with the skimage code here: https://stackoverflow.com/a/52443013/2012643
I think that we were just setting each contour pixel to 1 and leaving the rest black which led to gaps in the boundaries. However, cv2 method creates a line between the contour pixel and a certain offset point, for every contour pixel. This most probably removes the issue of gaps in the boundaries.